### PR TITLE
Search bar & Location Dropdown

### PR DESCRIPTION
**Pull request by AarCon**
This adds a search bar for the pokemon which then returns a list of locations that this mon can be found. Eventually we will have to standardize the names of locations in the coordinates.js so that this will accurately reflect the different areas on the map.

This PR also adds a location dropdown that updates with either selecting a option from the dropdown or clicking on the location.

The below videos show what is happening in the background especially with the pokemon selection

https://github.com/TeamLumi/luminescent-team-dev/assets/125856819/4a9aec6c-f29d-417d-bcd5-83558bd515ae

https://github.com/TeamLumi/luminescent-team-dev/assets/125856819/84fde974-4d2a-4cf0-8bd9-e49f08434ba0

### CODE REVIEW

File: src/components/Mapper/Mapper.jsx [hunk 1]
```diff
@@ -44,10 +45,18 @@ function getSelectedLocation(x, y) {
   return location[0];
 }
 
-export default function Mapper() {
+const canvasDimensions = {
+  width: 1244,
```
Comment (Denmark14) [inline src/components/Mapper/Mapper.jsx:65]: We'll need a better way to assign dimensions later. based on user's screen. Extract to config for now

File: src/components/Mapper/Mapper.jsx [hunk 2]
```diff
@@ -69,6 +78,10 @@ export default function Mapper() {
     setEncounterList(setAllEncounters(locationName))
   }, [swarm, radar, tod, incense, surfIncense, rod])
 
+  useEffect(() => {
+    setLocationList(getRoutesFromPokemonId(getPokemonIdFromName(pokemon)))
+  }, [pokemon])
+
   const handleTODChange = (event, nextView) => {
```
Comment (Denmark14) [inline src/components/Mapper/Mapper.jsx]: Acronyms are evil

File: src/components/Mapper/Mapper.jsx [hunk 3]
```diff
@@ -81,6 +94,10 @@ export default function Mapper() {
     callback(event.target.checked);
   };
 
+  const handlePokemonChange = (text) => {
```
Comment (Denmark14) [inline src/components/Mapper/Mapper.jsx]: What is text? What is the Pokemon change?

File: src/components/Mapper/Mapper.jsx [hunk 4]
```diff
@@ -174,7 +191,7 @@ export default function Mapper() {
       myCanvas.current.removeEventListener('click', handleClick);
       myCanvas.current.addEventListener('mousemove', handleMouseMove);
     };
-  }, [swarm, radar, tod, incense, surfIncense, rod])
+  }, [swarm, radar, tod, incense, surfIncense, rod, pokemon])
```
Comment (Denmark14) [inline src/components/Mapper/Mapper.jsx]: pokemon what? id or monsno?

**Review by Denmark14 — CHANGES_REQUESTED** [review]

**Comment by Denmark14** [comment]
I feel like showing a list of everywhere a Pokemon is available at once is beyond the scope of the Mapper. Last I checked, we did not support the highlighting of multiple areas as it would overflow items, etc. We should extract this component and move it to where we discussed previously on the dex, being where the held item component currently is.

File: src/components/Mapper/Mapper.jsx (hunk 1 above)
Comment (AarCon) [reply src/components/Mapper/Mapper.jsx:65]: It also means that we would need a way to dynamically assign the coordinates which sounds like a really difficult thing to do.

File: src/components/Mapper/Mapper.jsx (hunk 2 above)
Comment (AarCon) [reply src/components/Mapper/Mapper.jsx]: Changed every instance to TimeOfDay or timeOfDay

File: src/components/Mapper/Mapper.jsx (hunk 4 above)
Comment (AarCon) [reply src/components/Mapper/Mapper.jsx]: Changed all of the `pokemon` references to be `pokemonName` and derivatives of that.

**Comment by AarCon** [comment]
Showing the actual list of where a pokemon is available isn't the point of the pokemon dropdown, it's to highlight those areas. The video showing that was just a POC of what is happening in the background after choosing the pokemon. The user won't be able to see the actual list at all. And as for the highlighting, it's honestly really simple. All that it needs to do is, while it's drawing out the boundaries, to check if any of the `coordinates.name`s are in the location list which is returned from the `getRoutesFromPokemonId` function.

As for drawing the overlay, that is handled in the `feature/highlight_zones` branch like so:
```js
  const drawOverlay = (ctx) => {

    coordinates.forEach(coord => {
      // Draw zone outlines
      ctx.beginPath();
      ctx.moveTo(coord.x, coord.y);
      ctx.lineTo(coord.x + coord.w, coord.y);
      ctx.lineTo(coord.x + coord.w, coord.y + coord.h);
      ctx.lineTo(coord.x, coord.y + coord.h);
      ctx.closePath();
      if (locationList.includes(coord.name)) { // locationList is the list of locations you can find mons
        ctx.fillStyle = `rgba(${colors.enc.r}, ${colors.enc.g}, ${colors.enc.b}, ${colors.enc.a})`;
        ctx.fill();
      }
      if (hoveredZone === coord.name && hoveredZone !== locationName) {
        ctx.fillStyle = `rgba(${colors.hov.r}, ${colors.hov.g}, ${colors.hov.b}, ${colors.hov.a})`;
        ctx.fill();
      }
      if (locationName === coord.name) {
        ctx.fillStyle = `rgba(${colors.sel.r}, ${colors.sel.g}, ${colors.sel.b}, ${colors.sel.a})`;
        ctx.fill();
      }

      ctx.strokeStyle = 'rgba(0, 0, 0, 1)';
      ctx.lineWidth = hoveredZone === coord.name ? 2.3 : 1;
      ctx.stroke();
    });
  };
```
This function is run whenever the zone that the cursor is hovered over changes, on click, and whenever any of the encounter options changes. (we likely don't need to update that specific useEffect with encounter options tho) The way that it's changed is basically by scrapping the last iteration of the canvas and drawing a new one. I guess this could run into some performance issues if you're hovering over zones too quickly, but I don't imagine that will cause it to crash.

The main lines to look for are these ones where if the locationList includes any of the `coordinates.name`, then it will highlight them with the appropriate color. Unfortunately, I had to separate this from the function that actually generates this list because I didn't want to have to go through hooking the mapper function into the pokedex plugin again.
```js
      if (locationList.includes(coord.name)) { // locationList is the list of locations you can find mons
        ctx.fillStyle = `rgba(${colors.enc.r}, ${colors.enc.g}, ${colors.enc.b}, ${colors.enc.a})`;
        ctx.fill();
      }
```

File: src/components/Mapper/Mapper.jsx (hunk 1 above)
Comment (Denmark14) [reply src/components/Mapper/Mapper.jsx:65]: Fair point, not MVP worthy

**Review by Denmark14 — APPROVED** [review]

**Pull request merged**